src/compiler/checker.ts

```diff
@@ -6922,7 +6964,15 @@ namespace ts {
             return result & TypeFlags.PropagatingFlags;
         }
 
+        // This function replaces substitution types with their underlying type parameters. We erase when creating
+        // type references and type alias instantiations because subsitution types are no longer necessary once
+        // the type arguments have been validated against their corresponding type parameter constraints.

```

Comment:
Why does this only happen with type reference and type alias instantiations, but not something like index types? For example, I would expect that you'd need to erase the substitution for `T` here:

```ts
T extends any[] ? T[number] : never;
```

Reply:
Actually, it turns out that erasing substitution types when creating type references and type alias instantiations is a bit to aggressive, so I'm going to remove this. The right place to erase them is when the substitution type itself is instantiated (because whatever the instantiation provides will already be implied by the type that the instantiation maps to), and we already do this.